src/CameraOnvif.cpp

```diff
@@ -0,0 +1,408 @@
+#include "CameraOnvif.hpp"
+
+#include <gsoap/plugin/wsseapi.h>
+#include <gsoap/plugin/wsddapi.h>
+#include "onvif/soapDeviceBindingProxy.h"
+#include "onvif/soapImagingBindingProxy.h"
+#include "onvif/soapMediaBindingProxy.h"
+#include "onvif/soapPTZBindingProxy.h"
+#include "onvif/soapPullPointSubscriptionBindingProxy.h"
+#include "onvif/soapRemoteDiscoveryBindingProxy.h"
+#include "wsdd.nsmap"
+
+#include <iostream>
+
+using namespace std;
+using namespace camera_onvif;
+
+struct CameraOnvif::Private {
+    struct soap *soap = soap_new1(SOAP_XML_STRICT | SOAP_XML_CANONICAL | SOAP_C_UTFSTRING);
+    MediaBindingProxy *proxy_media;
+    ImagingBindingProxy *proxy_image;
+    _trt__GetVideoEncoderConfigurationOptionsResponse *video_options;
+}defined;
+
+CameraOnvif::CameraOnvif(/* args */) {
+    user = "admin";
+    pass = "camera01";
+    uri = "http://10.20.0.188/onvif/device_service";
+
+    this->init();
+}
+
+CameraOnvif::CameraOnvif(string user, string pass, string ip){
```

Comment:
Parameter initialization in C++ is done as initialization list:

~~~
CameraOnvif::CameraOnvif(string user, string pass, string ip)
    : user(user), pass(pass), ip(ip) {
}
~~~

This ensures that all the attributes are initialized before the constructor gets called (i.e. before the object itself is fully built), and e.g. that nothing will end up uninitialized in the destructor. This is not critical in this case (because strings), but could be for instance for a pointer. So, best practice is to initialize as much in the initializer list.

Reply:
I don't remember to see any tutorial explaining that...